docs/developer/design/authentication.rst

```diff
@@ -0,0 +1,210 @@
+
+Unexploited choices
+~~~~~~~~~~~~~~~~~~~
+
+* `Guard <https://github.com/appscode/guard/>`_
+
+A Kubernetes webhook authentication server by AppsCode, allowing you to
+log into your Kubernetes cluster by using various identity providers such as
+LDAP.
+
+* `ORY Hydra <https://github.com/ory/hydra/>`_
+
+It's an OpenID Connect provider optimized for low resource consumption.
+ORY Hydra is not an identity provider but it is able to connect to existing
+identity providers.
+
+Implementation Details
+----------------------
+
+Implementation steps(High level)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Iteration 1
+~~~~~~~~~~~
+
+- Using Salt, generate self-signed certificates needed for Dex deployment
+- Deploy Dex in MetalK8s from the official **Dex Charts** while making use of
+  the generated certificates above
+- Configure Kubernetes API server flags to use Dex
+- Expose Dex on the control plane using Ingress
+- Implement MetalK8s UI integration with Dex
+- Theme the Dex GUI to match MetalK8s UI specs
+
+Iteration 2
+~~~~~~~~~~~
+
+- Provide documentation on how to integrate with these external Identity
+  Providers
+
+Iteration 3
+~~~~~~~~~~~
+
+- Provide a CLI for user management and provisioning
+
+The following operations will be supported using the CLI tool:
+
+- Create users password
+- List existing passwords
+- Delete users password
+- Edit existing password
+- Revoke a users token
+
+The CLI tool will also be used to create MetalK8s dedicated roles as already
+specified in the requirements section of this document.
+(see high-level roles from the requirements document)
+
+Since the above mentioned operations can not be performed at the Dex ConfigMap
+level, using the Dex gRPC API is the way to go.
+
+Iteration 4
+~~~~~~~~~~~
+
+- Provide UI integration that performs similar CLI operations for user
+  management and provisioning.
+
+This means from the MetalK8s UI, a Cluster administrator should be able to do
+the following:
+
+- Create passwords for users
+- List existing passwords
+- Delete users password
+- Edit existing password
+- Revoke a users token
+
+.. note::
+
+   This iteration is completely optional for reasons being that the Identity
+   Provider from Dex acts as a fallback for Kubernetes Administrators who do
+   not want to use an external Identity Provider(mostly likely because they
+   have very small user store).
+
+Documentation
+-------------
+
+In the Operational Guide:
+
+* Document the predefined dex roles(Cluster Admin role, Solution Admin role,
+  Read Only role), their access levels and how to create them.
+* Document how to create users and the secrets associated to them.
+* Document how to integrate Dex with external Identity Provider(LDAP).
+
+
+Test Plan
+---------
+
+We could add some automated end-to-end tests for Dex user creation,
+and deletion using the CLI and then setup a mini-lab on scality cloud to try
+out the UI integration.
+
+Provide a virtualbox disk image of Microsoft server configured with
```

Comment:
We don't do VirtualBox in CI and such. Need to sync with the team that's done SAML integration for S3C on how to test this in CI.